File: backend/app/models/agent.py
"""Request and response models for the stock research agent API."""

from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, HttpUrl, field_validator

AgentRole = Literal["user", "assistant"]
ConfidenceLevel = Literal["verified", "document_based", "inferred", "unavailable"]


class AgentMessage(BaseModel):
    """One safe conversational message supplied to the agent."""

    model_config = ConfigDict(str_strip_whitespace=True)
    role: AgentRole
    content: str = Field(min_length=1, max_length=10_000)


class AgentChatRequest(BaseModel):
    """A new question plus bounded conversational context."""

    model_config = ConfigDict(str_strip_whitespace=True)
    message: str = Field(min_length=1, max_length=2_000)
    history: list[AgentMessage] = Field(default_factory=list, max_length=20)


class SourceReference(BaseModel):
    """A market-data or document source supporting the response."""

    label: str = Field(min_length=1, max_length=200)
    url: HttpUrl | None = None
    as_of: str | None = Field(default=None, max_length=50)


class ToolActivity(BaseModel):
    """User-safe summary of one tool execution."""

    name: str = Field(min_length=1, max_length=100)
    status: Literal["success", "error"]
    arguments: dict[str, Any] = Field(default_factory=dict)
    duration_ms: int | None = Field(default=None, ge=0)
    error: str | None = Field(default=None, max_length=500)

    @field_validator("error")
    @classmethod
    def require_error_for_failed_tool(cls, value: str | None, info: Any) -> str | None:
        if info.data.get("status") == "error" and not value:
            return "The tool could not complete the request."
        return value


class AgentChatResponse(BaseModel):
    """Evidence-aware answer returned by the agent endpoint."""

    answer: str = Field(min_length=1)
    trace_id: UUID
    provider: str = Field(min_length=1)
    model: str = Field(min_length=1)
    tools_used: list[str] = Field(default_factory=list)
    tool_activity: list[ToolActivity] = Field(default_factory=list)
    sources: list[SourceReference] = Field(default_factory=list)
    confidence: ConfidenceLevel = "inferred"
    warnings: list[str] = Field(default_factory=list)
    disclaimer: str = (
        "Educational research only; this is not personalized investment advice."
    )

    @field_validator("tools_used")
    @classmethod
    def unique_tools(cls, values: list[str]) -> list[str]:
        return list(dict.fromkeys(value.strip() for value in values if value.strip()))

File: frontend/types/agent.ts
export type AgentRole = "user" | "assistant";
export type ConfidenceLevel =
  | "verified"
  | "document_based"
  | "inferred"
  | "unavailable";
export type DataStatus =
  | "mock"
  | "end_of_day"
  | "delayed"
  | "real_time"
  | "unavailable";
export type ToolStatus = "running" | "success" | "error";

export type AgentHistoryMessage = {
  role: AgentRole;
  content: string;
};

export type AgentChatRequest = {
  message: string;
  history?: AgentHistoryMessage[];
};

export type ResearchSource = {
  label: string;
  url?: string | null;
  as_of?: string | null;
};

export type ToolExecution = {
  name: string;
  status: ToolStatus;
  arguments?: Record<string, unknown>;
  duration_ms?: number | null;
  error?: string | null;
};

/** Supports both the initial endpoint and the evidence-aware agent response. */
export type AgentChatResponse = {
  answer: string;
  trace_id: string;
  provider: string;
  model: string;
  tools_used: string[];
  tool_activity?: ToolExecution[];
  sources: Array<ResearchSource | string>;
  confidence?: ConfidenceLevel;
  data_status?: DataStatus;
  warning?: string;
  warnings?: string[];
  disclaimer?: string;
};

export type AgentChatMessage = {
  id: string;
  role: AgentRole;
  content: string;
  createdAt?: string;
  traceId?: string;
  confidence?: ConfidenceLevel;
  dataStatus?: DataStatus;
  tools?: ToolExecution[];
  sources?: ResearchSource[];
  warnings?: string[];
};

export type UseAgentChatOptions = {
  initialMessages?: AgentChatMessage[];
  maxHistory?: number;
};

export type UseAgentChatResult = {
  messages: AgentChatMessage[];
  isLoading: boolean;
  error: string | null;
  sendMessage: (message: string) => Promise<void>;
  retryLastMessage: () => Promise<void>;
  stop: () => void;
  clearMessages: () => void;
  clearError: () => void;
};
